storm-core/src/jvm/backtype/storm/blobstore/NimbusBlobStore.java

```diff
@@ -0,0 +1,408 @@
+    }
+    
+    @Override
+    public synchronized int available() {
+      return buffer == null ? 0 : (end - offset);
+    }
+
+    @Override
+    public long getFileLength() {
+        return beginBlobDownload.get_data_size();
+    }
+  }
+
+  public class NimbusUploadAtomicOutputStream extends AtomicOutputStream {
+    private String session;
+    private int maxChunkSize = 4096;
+    private String key;
+
+    public NimbusUploadAtomicOutputStream(String session, int bufferSize, String key) {
+      this.session = session;
+      this.maxChunkSize = bufferSize;
+      this.key = key;
+    }
+
+    @Override
+    public void cancel() throws IOException {
+      try {
+        synchronized(client) {
+          client.getClient().cancelBlobUpload(session);
+        }
+      } catch (TException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+      try {
+        synchronized(client) {
+          client.getClient().uploadBlobChunk(session, ByteBuffer.wrap(new byte[] {(byte)b}));
+        }
+      } catch (TException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    
+    @Override
+    public void write(byte []b) throws IOException {
+      write(b, 0, b.length);
+    }
+    
+    @Override
+    public void write(byte []b, int offset, int len) throws IOException {
+      try {
+        int end = offset + len;
+        for (int realOffset = offset; realOffset < end; realOffset += maxChunkSize) {
+          int realLen = Math.min(end - realOffset, maxChunkSize);
+          LOG.debug("Writing {} bytes of {} remaining",realLen,(end-realOffset));
+          synchronized(client) {
+            client.getClient().uploadBlobChunk(session, ByteBuffer.wrap(b, realOffset, realLen));
+          }
+        }
+      } catch (TException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    
+    @Override
+    public void close() throws IOException {
+      try {
+        synchronized(client) {
+          client.getClient().finishBlobUpload(session);
+          client.getClient().createStateInZookeeper(key);
+        }
+      } catch (TException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private NimbusClient client;
+  private int bufferSize = 4096;
+
+  @Override
+  public void prepare(Map conf) {
+    this.client = NimbusClient.getConfiguredClient(conf);
+    if (conf != null) {
+        this.bufferSize = Utils.getInt(conf.get(Config.STORM_BLOBSTORE_INPUTSTREAM_BUFFER_SIZE_BYTES), bufferSize);
+    }
+  }
+
+  @Override
+  protected AtomicOutputStream createBlobToExtend(String key, SettableBlobMeta meta)
+      throws AuthorizationException, KeyAlreadyExistsException {
+    try {
+      synchronized(client) {
+        return new NimbusUploadAtomicOutputStream(client.getClient().beginCreateBlob(key, meta), this.bufferSize, key);
+      }
+    } catch (AuthorizationException | KeyAlreadyExistsException exp) {
```

Comment:
Here Authorization and KeyAlreadyExists are TExceptions as per thrift. Therefore, we will not be able to catch specfic exceptions and log them higher up in the stack